Webclient

titeuf87 opened the pull request:
#### Brief overview of PR changes/additions
Adds an option dialogs to the webclient with some basic options

#### Motivation for adding to Evennia
Part of #614

#### Other info (issues closed, discussion etc)
People will need to update the javascript files/templates if they customised it.
I also did not haven the chance to test this under IE.

Griatch commented on evennia/server/inputfuncs.py:
```diff
@@ -421,3 +421,62 @@ def unmonitor(session, *args, **kwargs):
+    Called when the webclient options stored on the player changes.
+    Inform the interested clients of this change.
+    """
+    session = kwargs["session"]
```
This probably needs to be checked for existence before continuing (`if session: ...`); there are a few cases (such as abrupt client disconnect at *just* the right moment) where session may be `None`.

titeuf87 replied:
That's changed.

Griatch commented:
Overall this looks very good. I have a few usability points: 

- The option window does not close if clicking the option "cog wheel" icon a second time. It seems to be a untuitive thing to do. 
- There is no keyboard shortcut to close the option window (`escape` and/or `return` are the natural keys for a keyboard user to try). 
- The "popup window" notification type is maybe mis-named. "Popup-window" feels like a large new window should pop up, when it's really (in most OS:es) a small notification in a corner. Maybe "Popup notification" or simply just "Popup" would be a better name?
- While the functionality of "gag prompt from output" makes sense if one can test sending a prompt with `@py`, it is not clear from the text what it means, especially if one does not know how prompts interact with text normally. Maybe "Don't echo prompts to main text area" or something like that would be clearer? 
- On that note, considering it's likely the option pane expands over time, I wonder if not something like a hover-over option tooltip would not be important to have already at this early stage. If you don't want to add it for this PR, that's fine, but that would certainly be a feature request. 
- On a more fundamental level, the last point of #614 discusses an alternative to popups - the splitting of the main window into one or more panes. Multiple panes seems more suitable for a keyboard-dominated playflow than a popup that you may need to move around with the mouse. The option window should definitely remain a popup  but my thinking is that the usage of different window redirection in the future would aim towards something like this: 
  1. Split out the Client surface the way you want (either freely as a player or pre-set by the game dev, or both).
  2. Assign (from a list of game-supported tags) one or more tags to each pane to determine what should "go" into each window. 
  3. For example you could assign the "help" and "look" tags to the pane in top right of the client window. That's then where the webclient would direct `msg(text, {"window": "help"})` and `msg(text, {"window": "look"})` outputs. If a tag is not assigned, it just goes into the main window as usual. 

As said, several of these points are not something that need to be a apart of this PR, obviously, but it's something that should be considered (or discussed if people don't agree with the concepts) when it comes to designing the UI.

titeuf87 commented:
I'll fix the small usability issues this week.

Gag prompt from output: changing the wording is okay. But ideally I don't
really like this option as stock Evennia doesn't display any prompts at
all...
It'd be better if this is only visible if it's actually useful.

For the last point:
Seems like a good idea: having every single "output tag" (like help or
look) listed in the options, together with a combobox for each item to pick
where to render it: main output, popup, top-right pane, ...
One issue there is should all channels be grouped together? Or be able to
have channels in different panes?




On Sun, Jan 22, 2017 at 3:42 PM, Griatch <notifications@github.com> wrote:

> Overall this looks very good. I have a few usability points:
>
>    - The option window does not close if clicking the option "cog wheel"
>    icon a second time. It seems to be a untuitive thing to do.
>    - There is no keyboard shortcut to close the option window (escape
>    and/or return are the natural keys for a keyboard user to try).
>    - The "popup window" notification type is maybe mis-named.
>    "Popup-window" feels like a large new window should pop up, when it's
>    really (in most OS:es) a small notification in a corner. Maybe "Popup
>    notification" or simply just "Popup" would be a better name?
>    - While the functionality of "gag prompt from output" makes sense if
>    one can test sending a prompt with @py, it is not clear from the text
>    what it means, especially if one does not know how prompts interact with
>    text normally. Maybe "Don't echo prompts to main text area" or something
>    like that would be clearer?
>    - On that note, considering it's likely the option pane expands over
>    time, I wonder if not something like a hover-over option tooltip would not
>    be important to have already at this early stage. If you don't want to add
>    it for this PR, that's fine, but that would certainly be a feature request.
>    - On a more fundamental level, the last point of #614
>    <https://github.com/evennia/evennia/issues/614> discusses an
>    alternative to popups - the splitting of the main window into one or more
>    panes. Multiple panes seems more suitable for a keyboard-dominated playflow
>    than a popup that you may need to move around with the mouse. The option
>    window should definitely remain a popup but my thinking is that the usage
>    of different window redirection in the future would aim towards something
>    like this:
>       1. Split out the Client surface the way you want (either freely as
>       a player or pre-set by the game dev, or both).
>       2. Assign (from a list of game-supported tags) one or more tags to
>       each pane to determine what should "go" into each window.
>       3. For example you could assign the "help" and "look" tags to the
>       pane in top right of the client window. That's then where the webclient
>       would direct msg(text, {"window": "help"}) and msg(text, {"window":
>       "look"}) outputs. If a tag is not assigned, it just goes into the
>       main window as usual.
>
> As said, several of these points are not something that need to be a apart
> of this PR, obviously, but it's something that should be considered (or
> discussed if people don't agree with the concepts) when it comes to
> designing the UI.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/evennia/evennia/pull/1172#issuecomment-274334864>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACQy8prwECRRBHKovLnTMVKGAo4jYxruks5rU2rngaJpZM4LqXAD>
> .
>



-- 
Simon Vermeersch

Griatch commented:
@titeuf87 To make my thoughts on the future webclient GUI a little clearer (also in my own mind) I made a new brainstorming whitepage for it here, check it out and give feedback as needed: https://github.com/evennia/evennia/wiki/Webclient-brainstorm

Griatch commented:
I think there are still some usability things to handle on this before it can merge.

titeuf87 commented:
I've changed the various usability remarks.

For the help popup: this clashes with the recent changes to show help text using EvMore. I changed it so that EvMore is not used for the webclient when help is shown in a separate popup. Not sure it is ideal though...

If there are no more issues with this I can start working on the fancy pane layout :)

Griatch commented:
This looks very good now; the only remaning thing is that EvMore + help pane thing. I don't think it's a good idea to turn off EvMore on the Server level based on the protocol used - this sets a bad precedence considering the Server should be as protocol-agnostic as possible. But I understand that a server-side pager like EvMore could be hard for the popup help pane to handle as it's written since the pane doesn't accept input. 
Supposedly, if the pane system is coming, the help entry will not need a popup. But if one kept it, it sounds like the popup needs to made aware that it should accept input, at least the ones needed by EvMore (in this case). If so, maybe EvMore could tag the outgoing OOB message to make the pane aware it is an EvMore situation? I imagine this could be useful for other things too. Opinions?

titeuf87 commented:
In that case...if EvMore is used and the text it renders ends up in a separate pane, is it still useful to actually have the paging functionality?
That pane will have scrollbars from the browser and could be setup to not automatically scroll at the bottom so that when a player opens a big help file it starts at the beginning and they can scroll.
And it is possible to add buttons/shortcut keys to easily scroll using the keyboard (though not sure how the keyboard shortcuts would work if there is more than one pane open that can be scrolled).

If text gets rendered to the main output then EvMore can work like it does now as there I can imagine that paging makes more sense.

Griatch commented:
Hm, good point about the scroll bars. The issue is though the EvMore happens on the server level - the client cannot handle it differently because evmore simply doesn't supply it with any more text at a time than evmore allows. 
I pondered a few solutions to let the client handle an evmore call, but you run into issues with the evmore cmdset at any rate. Ok, it may well be best to handle this at the Server level the way you do, at least while we are using a popup. I'll look it over and merge it.

Griatch commented:
Ok merged with a rebase. I renamed the `msg(text, {"window": "help"}` to `msg(text, {"type":"help"}` which seems more future-proof and generic as a kwarg name. Thanks for your good work with this!
